pandas/core/groupby/groupby.py

```diff
@@ -847,43 +857,70 @@ def _transform_should_cast(self, func_nm):
 
     def _cython_transform(self, how, numeric_only=True, **kwargs):
         output = collections.OrderedDict()
-        for name, obj in self._iterate_slices():
+        names = []  # type: List[Hashable]
+        for index, (name, obj) in enumerate(self._iterate_slices()):
             is_numeric = is_numeric_dtype(obj.dtype)
             if numeric_only and not is_numeric:
                 continue
 
             try:
-                result, names = self.grouper.transform(obj.values, how, **kwargs)
+                result, _ = self.grouper.transform(obj.values, how, **kwargs)
             except NotImplementedError:
                 continue
+
             if self._transform_should_cast(how):
-                output[name] = self._try_cast(result, obj)
-            else:
-                output[name] = result
+                result = self._try_cast(result, obj)
+
+            output[index] = result
+            names.append(name)
 
         if len(output) == 0:
             raise DataError("No numeric types to aggregate")
 
         return self._wrap_transformed_output(output, names)
 
-    def _wrap_aggregated_output(self, output, names=None):
+    def _wrap_aggregated_output(
+        self, output: Mapping[int, np.ndarray], names: List[Hashable]
+    ):
         raise AbstractMethodError(self)
 
-    def _wrap_transformed_output(self, output, names=None):
+    def _wrap_transformed_output(
+        self, output: Mapping[int, np.ndarray], names: List[Hashable]
+    ):
         raise AbstractMethodError(self)
 
     def _wrap_applied_output(self, keys, values, not_indexed_same=False):
         raise AbstractMethodError(self)
 
     def _cython_agg_general(self, how, alt=None, numeric_only=True, min_count=-1):
-        output = {}
+        output = collections.OrderedDict()
```

Comment:
does OrderedDict make a difference here?  we do something similar in ops with just a regular dict.

Reply:
I don't totally get it, but trust you on this

Reply:
Well...I'm hoping this becomes clearer so if not I would certainly like your feedback.

To illustrate, let's say we have a DataFrame that looks like:

```python
df = pd.DataFrame([[1, 1]], columns=["a", "b"])
```

And then do

```python
df.groupby(["x"]).any()
```

This hits the loop in `_get_cythonized_result`. The loop iterates over the DataFrame via `_iterate_slices`, which essentially goes by column. Within each iteration, the function gets applied to the columns values, and an entry is added to the `output` dict where the key goes from 0...len(columns) - 1. At the same time, the column name(s) is added as an entry to `names`

So ultimately that looks like:

```python
>>> output
{0: array([ True]), 1: array([ True])}
>> names
['a', 'b']
```

The wrapping functions then accept `output` and `names`, constructs a DataFrame from `output` which naturally would show as

```python
>>> pd.DataFrame(output)
      0     1
x  True  True
```

And then layers the names on top of that in order:

```python
      a     b
x  True  True
```

Obviously it would be great if `output` and `names` weren't too separate entities and we could just use the column label as the key to `output`, but that doesn't work when you have duplicated column labels.

An alternate approach which I tried at first didn't involve a dictionary to store the results by column, but this hit limitations when trying to build the result back up, namely because:

  - _You can't use one homogeneous ndarray_ because you'd lose dtype information, so maybe you'd then try
  - Building perhaps a list of ndarrays that you could concat together, but `pd.concat` cannot piece together ndarrays, so lastly you could try
  - Storing a list of `pd.Series` and contacting them at the end, but the performance of that lagged this by about 15-20%

Hope that helps. If not let me know and I can try to explain some other way

Reply:
In dispatch_to_series we do something like:

```
output = {}
for i in range(len(columns)):
    output[i] = func(df.iloc[:, i])

result = DataFrame(output)
result.columns = columns
```

In that context, having `output` be a dict and not an `OrderedDict` is fine.  The thing I'm not getting is what is different here that makes an OrderedDict necessary.

Reply:
Just guarding against dicts not being ordered in Py35. These can easily become dicts when we drop Py35 over the next few days